lightning/restore/restore.go

```diff
@@ -1367,154 +1519,31 @@ func (cr *chunkRestore) restore(
-			start := time.Now()
-			dataStream, err := dataEngine.NewWriteStream(ctx)
-			if err != nil {
-				deliverCompleteCh <- errors.Trace(err)
-				return
-			}
-			indexStream, err := indexEngine.NewWriteStream(ctx)
-			if err != nil {
-				deliverCompleteCh <- errors.Trace(err)
-				return
-			}
-			// class kvs
-			for _, k := range b.totalKVs {
-				if k.Key[tablecodec.TableSplitKeyLen+1] == 'r' {
-					dataKVs = append(dataKVs, k)
-				} else {
-					indexKVs = append(indexKVs, k)
-				}
-			}
-			b.totalKVs = nil
-
-			for _, kvs := range splitIntoDeliveryStreams(dataKVs, maxDeliverBytes) {
-				if e := dataStream.Put(kvs); e != nil {
-					if err != nil {
-						common.AppLogger.Warnf("failed to put write stream: %s", e.Error())
-					} else {
-						err = e
-					}
-				}
-			}
-			for _, kvs := range splitIntoDeliveryStreams(indexKVs, maxDeliverBytes) {
-				if e := indexStream.Put(kvs); e != nil {
-					if err != nil {
-						common.AppLogger.Warnf("failed to put write stream: %s", e.Error())
-					} else {
-						err = e
-					}
-				}
-			}
-			dataKVs = dataKVs[:0]
-			indexKVs = indexKVs[:0]
-
-			block.cond.Signal()
-			if e := dataStream.Close(); e != nil {
-				if err != nil {
-					common.AppLogger.Warnf("[%s:%d] failed to close write stream: %s", t.tableName, engineID, e.Error())
-				} else {
-					err = e
-				}
-			}
-			if e := indexStream.Close(); e != nil {
-				if err != nil {
-					common.AppLogger.Warnf("[%s:%d] failed to close write stream: %s", t.tableName, engineID, e.Error())
-				} else {
-					err = e
-				}
-			}
-			deliverDur := time.Since(start)
-			deliverTotalDur += deliverDur
-			metric.BlockDeliverSecondsHistogram.Observe(deliverDur.Seconds())
-			metric.BlockDeliverBytesHistogram.Observe(float64(b.localChecksum.SumSize()))
-
-			if err != nil {
-				if !common.IsContextCanceledError(err) {
-					common.AppLogger.Errorf("[%s:%d] kv deliver failed = %v", t.tableName, engineID, err)
-				}
-				// TODO : retry ~
-				deliverCompleteCh <- errors.Trace(err)
-				return
-			}
-
-			// Update the table, and save a checkpoint.
-			// (the write to the importer is effective immediately, thus update these here)
-			cr.chunk.Checksum.Add(&b.localChecksum)
-			cr.chunk.Chunk.Offset = b.chunkOffset
-			cr.chunk.Chunk.PrevRowIDMax = b.chunkRowID
-			rc.saveCpCh <- saveCp{
-				tableName: t.tableName,
-				merger: &RebaseCheckpointMerger{
-					AllocBase: t.alloc.Base() + 1,
-				},
-			}
-			rc.saveCpCh <- saveCp{
-				tableName: t.tableName,
-				merger: &ChunkCheckpointMerger{
-					EngineID: engineID,
-					Key:      cr.chunk.Key,
-					Checksum: cr.chunk.Checksum,
-					Pos:      cr.chunk.Chunk.Offset,
-					RowID:    cr.chunk.Chunk.PrevRowIDMax,
-				},
-			}
-		}
+		dur, err := cr.deliverLoop(ctx, kvsCh, t, engineID, dataEngine, indexEngine, rc)
+		deliverCompleteCh <- deliverResult{dur, err}
 	}()
 
-	var buffer bytes.Buffer
+	defer close(kvsCh)
+
+	initializedColumns := false
+outside:
 	for {
 		select {
```

Comment:
We should check `deliverCompleteCh` here, think about the below scenario:

1. `cr.deliverLoop` occurred an error in any branch L1453/L1461/L1463
2. will send a result `deliverResult{dur, err}` to `deliverCompleteCh`
3. nothing will be received from `kvsCh` because of `cr.deliverLoop` is returned
4. L1453 will be blocked if the channel is full.

Oops, dead lock.

Reply:
we should check it while we need send data into the input channel